index.js

```diff
@@ -443,6 +459,228 @@ Command.prototype.allowUnknownOption = function(arg) {
   return this;
 };
 
+/**
+ * Define completionRules which will later be used by autocomplete to generate appropriate response
+ *
+ * @param {Object} completion rules
+ * @api public
+ */
+
+Command.prototype.complete = function(rules) {
+  // merge options
+  // this should ensure this._completionRules are always in shape
+  if (rules.options) {
+    this._completionRules.options = rules.options;
+  }
+
+  // support both arguments or args as key
+  if (rules.arguments) {
+    this._completionRules.args = rules.arguments;
+  } else if (rules.args) {
+    this._completionRules.args = rules.args;
+  }
+
+  return this;
+};
+
+/**
+ * Handle autocomplete if command args starts with special options.
+ * It will exit current process after successful processing.
+ *
+ * @param {Array} process.argv
+ */
+
+Command.prototype.autocomplete = function(argv) {
+  var RESERVED_STARTING_KEYWORDS = [
+    '--completion',
+    '--completion-fish',
+    '--compzsh',
+    '--compbash',
+    '--compfish'
+  ];
+  var firstArg = argv[2];
+
+  if (RESERVED_STARTING_KEYWORDS.includes(firstArg)) {
```

Comment:
I worry here about breaking change if I understand correctly now reserved flags will be handled internally by commander and breaks users' code which use one/some of those options before.

Reply:
Added a guarding condition statement around `this.autocomplete()` call in parse function. only if at least one rule has been defined we'll run into argv checks.

In this case, we can assume that the user is aware of the completion feature and will solve the confliction of test flags if there's any.